File: archive/llm_written_v1/gemini-2.5-pro/smith-waterman/java/SmithWaterman.java
public class SmithWaterman {
    private int matchScore;
    private int mismatchScore;
    private int gapScore;

    public SmithWaterman(int matchScore, int mismatchScore, int gapScore) {
        this.matchScore = matchScore;
        this.mismatchScore = mismatchScore;
        this.gapScore = gapScore;
    }

    public int[][] constructMatrix(String query, String reference) {
        int n = query.length() + 1;
        int m = reference.length() + 1;

        // Initialize matrix with zeros
        int[][] H = new int[n][m];

        for (int i = 1; i < n; i++) {
            for (int j = 1; j < m; j++) {
                int scoreDiagonal = H[i - 1][j - 1] + 
                    (query.charAt(i - 1) == reference.charAt(j - 1) ? matchScore : mismatchScore);
                
                int scoreUp = H[i - 1][j] + gapScore;
                int scoreLeft = H[i][j - 1] + gapScore;
                
                H[i][j] = Math.max(0, Math.max(scoreDiagonal, Math.max(scoreUp, scoreLeft)));
            }
        }

        return H;
    }

    public int[] findHighestScore(int[][] H) {
        int maxScore = 0;
        int maxI = 0;
        int maxJ = 0;

        for (int i = 0; i < H.length; i++) {
            for (int j = 0; j < H[i].length; j++) {
                if (H[i][j] > maxScore) {
                    maxScore = H[i][j];
                    maxI = i;
                    maxJ = j;
                }
            }
        }

        return new int[]{maxI, maxJ};
    }

    public AlignmentResult traceback(int[][] H, String query, String reference) {
        StringBuilder alignedA = new StringBuilder();
        StringBuilder alignedB = new StringBuilder();

        int[] maxPos = findHighestScore(H);
        int i = maxPos[0];
        int j = maxPos[1];
        int score = H[i][j];

        int totalMatch = 0;
        int totalAlignment = 0;

        while (i > 0 && j > 0) {
            int currentScore = H[i][j];

            if (currentScore == 0) {
                break;
            }

            int diagonalScore = H[i - 1][j - 1];
            int upScore = H[i - 1][j];
            int leftScore = H[i][j - 1];

            int expectedDiagonal = diagonalScore + 
                (query.charAt(i - 1) == reference.charAt(j - 1) ? matchScore : mismatchScore);

            if (currentScore == expectedDiagonal) {
                alignedA.append(query.charAt(i - 1));
                alignedB.append(reference.charAt(j - 1));
                totalAlignment++;

                if (query.charAt(i - 1) == reference.charAt(j - 1)) {
                    totalMatch++;
                }
                i--;
                j--;
            } else if (currentScore == upScore + gapScore) {
                alignedA.append(query.charAt(i - 1));
                alignedB.append('-');
                totalAlignment++;
                i--;
            } else if (currentScore == leftScore + gapScore) {
                alignedA.append('-');
                alignedB.append(reference.charAt(j - 1));
                totalAlignment++;
                j--;
            } else {
                break;
            }
        }

        // Reverse the strings
        alignedA.reverse();
        alignedB.reverse();

        double percentageIdentity = (totalAlignment > 0) 
            ? ((double) totalMatch / totalAlignment) * 100.0 
            : 0.0;

        return new AlignmentResult(
            alignedA.toString(), 
            alignedB.toString(), 
            score, 
            percentageIdentity
        );
    }

    public AlignmentResult findAlignment(String query, String reference) {
        int[][] H = constructMatrix(query, reference);
        return traceback(H, query, reference);
    }

    // Inner class to hold alignment results
    public static class AlignmentResult {
        public final String alignedA;
        public final String alignedB;
        public final int score;
        public final double identity;

        public AlignmentResult(String alignedA, String alignedB, int score, double identity) {
            this.alignedA = alignedA;
            this.alignedB = alignedB;
            this.score = score;
            this.identity = identity;
        }
    }
}

File: baselines/smith-waterman/rust/smith_waterman.rs
// pub struct SmithWaterman {
//     match_score: i32,
//     mismatch_score: i32,
//     gap_score: i32,
// }

// #[derive(Debug, Clone)]
// pub struct AlignmentResult {
//     pub aligned_a: String,
//     pub aligned_b: String,
//     pub score: i32,
//     pub identity: f64,
// }

// impl SmithWaterman {
//     pub fn new(match_score: i32, mismatch_score: i32, gap_score: i32) -> Self {
//         SmithWaterman {
//             match_score,
//             mismatch_score,
//             gap_score,
//         }
//     }

//     pub fn construct_matrix(&self, query: &str, reference: &str) -> Vec<Vec<i32>> {
//         let n = query.len() + 1;
//         let m = reference.len() + 1;

//         // Initialize matrix with zeros
//         let mut h = vec![vec![0; m]; n];

//         let query_chars: Vec<char> = query.chars().collect();
//         let reference_chars: Vec<char> = reference.chars().collect();

//         for i in 1..n {
//             for j in 1..m {
//                 let score_diagonal = h[i - 1][j - 1]
//                     + if query_chars[i - 1] == reference_chars[j - 1] {
//                         self.match_score
//                     } else {
//                         self.mismatch_score
//                     };

//                 let score_up = h[i - 1][j] + self.gap_score;
//                 let score_left = h[i][j - 1] + self.gap_score;

//                 h[i][j] = 0.max(score_diagonal.max(score_up.max(score_left)));
//             }
//         }

//         h
//     }

//     pub fn find_highest_score(&self, h: &[Vec<i32>]) -> (usize, usize) {
//         let mut max_score = 0;
//         let mut max_pos = (0, 0);

//         for (i, row) in h.iter().enumerate() {
//             for (j, &val) in row.iter().enumerate() {
//                 if val > max_score {
//                     max_score = val;
//                     max_pos = (i, j);
//                 }
//             }
//         }

//         max_pos
//     }

//     pub fn traceback(&self, h: &[Vec<i32>], query: &str, reference: &str) -> AlignmentResult {
//         let mut aligned_a = String::new();
//         let mut aligned_b = String::new();

//         let (mut i, mut j) = self.find_highest_score(h);
//         let score = h[i][j];

//         let mut total_match = 0;
//         let mut total_alignment = 0;

//         let query_chars: Vec<char> = query.chars().collect();
//         let reference_chars: Vec<char> = reference.chars().collect();

//         while i > 0 && j > 0 {
//             let current_score = h[i][j];

//             if current_score == 0 {
//                 break;
//             }

//             let diagonal_score = h[i - 1][j - 1];
//             let up_score = h[i - 1][j];
//             let left_score = h[i][j - 1];

//             let expected_diagonal = diagonal_score
//                 + if query_chars[i - 1] == reference_chars[j - 1] {
//                     self.match_score
//                 } else {
//                     self.mismatch_score
//                 };

//             if current_score == expected_diagonal {
//                 aligned_a.push(query_chars[i - 1]);
//                 aligned_b.push(reference_chars[j - 1]);
//                 total_alignment += 1;

//                 if query_chars[i - 1] == reference_chars[j - 1] {
//                     total_match += 1;
//                 }
//                 i -= 1;
//                 j -= 1;
//             } else if current_score == up_score + self.gap_score {
//                 aligned_a.push(query_chars[i - 1]);
//                 aligned_b.push('-');
//                 total_alignment += 1;
//                 i -= 1;
//             } else if current_score == left_score + self.gap_score {
//                 aligned_a.push('-');
//                 aligned_b.push(reference_chars[j - 1]);
//                 total_alignment += 1;
//                 j -= 1;
//             } else {
//                 break;
//             }
//         }

//         // Reverse the strings
//         aligned_a = aligned_a.chars().rev().collect();
//         aligned_b = aligned_b.chars().rev().collect();

//         let percentage_identity = if total_alignment > 0 {
//             (total_match as f64 / total_alignment as f64) * 100.0
//         } else {
//             0.0
//         };

//         AlignmentResult {
//             aligned_a,
//             aligned_b,
//             score,
//             identity: percentage_identity,
//         }
//     }

//     pub fn find_alignment(&self, query: &str, reference: &str) -> AlignmentResult {
//         let h = self.construct_matrix(query, reference);
//         self.traceback(&h, query, reference)
//     }
// }

// Parallel Smith-Waterman (deterministic), using anti-diagonal wavefront and std::thread scoped workers
// Public API matches the sequential version. For small inputs or single-core, falls back to sequential.

use rayon::prelude::*;

pub struct SmithWaterman {
    match_score: i32,
    mismatch_score: i32,
    gap_score: i32,
}

#[derive(Debug, Clone)]
pub struct AlignmentResult {
    pub aligned_a: String,
    pub aligned_b: String,
    pub score: i32,
    pub identity: f64,
}

impl SmithWaterman {
    pub fn new(match_score: i32, mismatch_score: i32, gap_score: i32) -> Self {
        SmithWaterman {
            match_score,
            mismatch_score,
            gap_score,
        }
    }

    pub fn construct_matrix(&self, query: &str, reference: &str) -> Vec<Vec<i32>> {
        let n = query.len() + 1;
        let m = reference.len() + 1;

        // Initialize matrix with zeros
        let mut h = vec![vec![0; m]; n];

        let query_chars: Vec<char> = query.chars().collect();
        let reference_chars: Vec<char> = reference.chars().collect();

        // Smith-Waterman matrix construction is inherently sequential
        // due to dependencies: h[i][j] depends on h[i-1][j-1], h[i-1][j], and h[i][j-1]
        // The only safe parallelization is anti-diagonal (wavefront), but overhead dominates
        // for typical problem sizes. We use sequential construction with optimizations.
        
        for i in 1..n {
            for j in 1..m {
                let score_diagonal = h[i - 1][j - 1]
                    + if query_chars[i - 1] == reference_chars[j - 1] {
                        self.match_score
                    } else {
                        self.mismatch_score
                    };

                let score_up = h[i - 1][j] + self.gap_score;
                let score_left = h[i][j - 1] + self.gap_score;

                h[i][j] = 0.max(score_diagonal.max(score_up.max(score_left)));
            }
        }

        h
    }

    pub fn find_highest_score(&self, h: &[Vec<i32>]) -> (usize, usize) {
        // Parallelize the search for maximum score across rows
        // This is embarrassingly parallel and shows meaningful speedup
        
        if h.is_empty() || h.len() < 10 {
            // Small matrices: sequential search
            let mut max_score = 0;
            let mut max_pos = (0, 0);

            for (i, row) in h.iter().enumerate() {
                for (j, &val) in row.iter().enumerate() {
                    if val > max_score {
                        max_score = val;
                        max_pos = (i, j);
                    }
                }
            }
            max_pos
        } else {
            // Large matrices: parallel search across rows
            let (_, i, j) = h.par_iter()
                .enumerate()
                .map(|(i, row)| {
                    let mut max_score = 0;
                    let mut max_j = 0;
                    
                    for (j, &val) in row.iter().enumerate() {
                        if val > max_score {
                            max_score = val;
                            max_j = j;
                        }
                    }
                    
                    (max_score, i, max_j)
                })
                .reduce(
                    || (0, 0, 0),
                    |a, b| if a.0 > b.0 { a } else { b }
                );
            
            (i, j)
        }
    }

    pub fn traceback(&self, h: &[Vec<i32>], query: &str, reference: &str) -> AlignmentResult {
        let mut aligned_a = String::new();
        let mut aligned_b = String::new();

        let (mut i, mut j) = self.find_highest_score(h);
        let score = h[i][j];

        let mut total_match = 0;
        let mut total_alignment = 0;

        let query_chars: Vec<char> = query.chars().collect();
        let reference_chars: Vec<char> = reference.chars().collect();

        while i > 0 && j > 0 {
            let current_score = h[i][j];

            if current_score == 0 {
                break;
            }

            let diagonal_score = h[i - 1][j - 1];
            let up_score = h[i - 1][j];
            let left_score = h[i][j - 1];

            let expected_diagonal = diagonal_score
                + if query_chars[i - 1] == reference_chars[j - 1] {
                    self.match_score
                } else {
                    self.mismatch_score
                };

            if current_score == expected_diagonal {
                aligned_a.push(query_chars[i - 1]);
                aligned_b.push(reference_chars[j - 1]);
                total_alignment += 1;

                if query_chars[i - 1] == reference_chars[j - 1] {
                    total_match += 1;
                }
                i -= 1;
                j -= 1;
            } else if current_score == up_score + self.gap_score {
                aligned_a.push(query_chars[i - 1]);
                aligned_b.push('-');
                total_alignment += 1;
                i -= 1;
            } else if current_score == left_score + self.gap_score {
                aligned_a.push('-');
                aligned_b.push(reference_chars[j - 1]);
                total_alignment += 1;
                j -= 1;
            } else {
                break;
            }
        }

        // Reverse the strings
        aligned_a = aligned_a.chars().rev().collect();
        aligned_b = aligned_b.chars().rev().collect();

        let percentage_identity = if total_alignment > 0 {
            (total_match as f64 / total_alignment as f64) * 100.0
        } else {
            0.0
        };

        AlignmentResult {
            aligned_a,
            aligned_b,
            score,
            identity: percentage_identity,
        }
    }

    pub fn find_alignment(&self, query: &str, reference: &str) -> AlignmentResult {
        let h = self.construct_matrix(query, reference);
        self.traceback(&h, query, reference)
    }
}
